### PR TITLE
DAOS-6159 test: Run fault injection tests in parallel and prepare for IL testing.

### PR DESCRIPTION
**Pull request by ashleypittman**
Update to the fault injection testing to run multiple instances in
parallel, up to the number of cores per node.  This dramatically
increases the speed of testing, and cuts down time-to-result.

Add the ability to run a read through the interception library under
fault injection, but do not enable this yet as it still causes a number
of issues.

Resolve the worst/easiest of the issues with the interception library
and the array object code.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
This decreases the NLT runtime from 7 minutes, to 3.5 minutes, although some of that will be lost when the additional testing is enabled.  The overall log files are similar in size and number which is to be expected, although some of the names have changed.  I've verified the memcheck re-runs are functioning correctly.

This PR adds support for running the daos client under fault injection, which itself exposes a number of errors, some are fixed in this PR but many remain so the new testing is turned off still, and will likely require someone more familiar with the code to resolve the remaining issues in the new year.

File: src/tests/ftest/cart/util/cart_logtest.py [hunk 1]
```diff
@@ -631,10 +641,11 @@ def _check_pid_from_log_file(self, pid, abort_on_warning,
...
 
-        print("Pid {}, {} lines total, {} trace ({:.2f}%)".format(pid,
-                                                                  total_lines,
-                                                                  trace_lines,
-                                                                  p_trace))
+        if not self.quiet:
+            print("Pid {}, {} lines total, {} trace ({:.2f}%)".format(pid,
+                                                                      total_lines,
+                                                                      trace_lines,
+                                                                      p_trace))
```
Comment (phender) [inline src/tests/ftest/cart/util/cart_logtest.py:648]: To avoid the line too-long pylint error you could use a [hanging indent](https://www.python.org/dev/peps/pep-0008/#indentation):
```suggestion
        if not self.quiet:
            print("Pid {}, {} lines total, {} trace ({:.2f}%)".format(
                pid, total_lines, trace_lines, p_trace))
```

Comment (ashleypittman) [reply src/tests/ftest/cart/util/cart_logtest.py:648]: True, although I don't really think that's much easier to read.  Short line lengths don't really work with python string formatting, but I'll add this change next time I have to rebuild.

File: src/client/array/dc_array.c [hunk 2]
```diff
@@ -1708,8 +1711,8 @@ dc_array_io(daos_handle_t array_oh, daos_handle_t th,
 					      tse_task2sched(task),
 					      0, NULL, &io_task);
 			if (rc != 0) {
-				D_ERROR("Fetch dkey "DF_U64" failed (%d)\n",
-					params->dkey_val, rc);
+				D_ERROR("Fetch dkey " DF_U64 " failed " DF_RC "\n",
```
Comment (liuxuezhao) [inline src/client/array/dc_array.c:1714]: I think better to fix the checkpatch script as "DF_U64" such usage is reasonable.

Comment (ashleypittman) [reply src/client/array/dc_array.c]: We've got code which uses both styles, but the majority of this file seems to be the shorted one so I've updated it.

This came up in the chat recently, we should probably decide to standardise on no spaces and update checkpatch as you say.

**Comment by ashleypittman** [comment]
As it stands this ticket does several things

- Reworks the FI testing to run in parallel significantly reducing runtime.
- Adds but not enables code to enable testing of the Interception library
- Fixes one deadlock, one double-free and several logging/memory leaks in the array code etc.

There are two issues with landing this PR however.

- If the IL-under-FI test is enabled then the server logs are too big, so the code needs to restart the server without logging.  I have other PRs that add this already
- There are many errors remaining in the client code, mostly seemingly in the tse and array code as shown here https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3902/18/VM_test/new/#issuesContent

I propose essentially landing this PR as-is because the contents are useful, then we create two PRs to address the above issues.  The second one probably needs to be done by someone more familiar with the array code than I am.  Obviously landing this would require the commit message/header to be update, I can do this if others are in agreement as to the way forward.

File: src/object/cli_obj.c [hunk 3]
```diff
@@ -1325,6 +1325,7 @@ dc_obj_open(tse_task_t *task)
 	obj_hdl_link(obj);
 	*args->oh = obj_ptr2hdl(obj);
 	obj_decref(obj);
+err:
```
Comment (mchaarawi) [inline src/object/cli_obj.c:1328]: should call this out instead of err.

File: src/common/fail_loc.c [hunk 4]
```diff
@@ -179,12 +179,19 @@ daos_fail_init(void)
 
 	rc = d_fault_inject_init();
 	if (rc != 0 && rc != -DER_NOSYS)
-		return rc;
+		D_GOTO(out, rc);
```
Comment (mchaarawi) [inline src/common/fail_loc.c:182]: unnecessary change

File: src/common/fail_loc.c [hunk 5]
```diff
@@ -179,12 +179,19 @@ daos_fail_init(void)
...
 
+	/* Log, but no not propagate error on registering the fault as this
+	 * leads to incorrect reference counts in daos_debug_init() and
+	 * resulting resource leaks.
+	 */
 	rc = d_fault_attr_set(DAOS_FAIL_UNIT_TEST_GROUP, attr);
-	if (rc)
-		d_fault_inject_fini();
+	if (rc != 0)
```
Comment (mchaarawi) [inline src/common/fail_loc.c:189]: we are not supposed to call d_fault_inject_fini()??
it was init earlier in the function.

Comment (ashleypittman) [reply src/common/fail_loc.c:189]: No.  This code is delicate.  From memory daos_fail_init() is ref counted and called from daos_debug_init() and daos_init(), if you call d_fault_inject_fini() here then it's properly closed but daos_debug_init() ignores the error, then daos_init() is called and re-initiaises it, at which point it works, so the first iteration of the fault injection test injects one fault (in fault_attr_set()), but then forgets it's supposed to forget one fault and carries on as normal.

I can do further testing to confirm but getting this startup code to work took a few iterations and this change was required.

Comment (ashleypittman) [reply src/common/fail_loc.c:189]: If you call d_fault_inject_fini() here then it's gets stuck in the following stack trace:
`#0  0x00007f14f1975585 in pthread_spin_lock () from /lib64/libpthread.so.0
#1  0x00007f14f043b578 in d_should_fail (fault_attr=0x24f7ee0) at src/gurt/fault_inject.c:659
#2  0x00007f14f042be25 in d_hash_table_create_inplace (feats=feats@entry=1, bits=bits@entry=13, priv=priv@entry=0x0, hops=hops@entry=0x7f14f06567e0 <fa_table_ops>, htable=htable@entry=0x7f14f0656ca0 <d_fi_gdata+64>) at src/gurt/hash.c:783
#3  0x00007f14f0439eec in d_fi_gdata_init () at src/gurt/fault_inject.c:438
#4  0x00007f14f196f20b in __pthread_once_slow () from /lib64/libpthread.so.0
#5  0x00007f14f043db5c in d_fault_inject_init () at src/gurt/fault_inject.c:477
#6  0x00007f14f06be56d in crt_init_opt (grpid=grpid@entry=0x0, flags=flags@entry=0, opt=0x0) at src/cart/crt_init.c:314
#7  0x00007f14f16509bb in daos_eq_lib_init () at src/client/api/event.c:90
#8  0x00007f14f16557c8 in daos_init () at src/client/api/init.c:184
#9  0x00007f14f27d703a in check_ioctl_on_open (fd=fd@entry=5, entry=entry@entry=0x7ffce0493e40, flags=flags@entry=0, status=<optimized out>) at src/client/dfuse/il/int_posix.c:601
#10 0x00007f14f27d81f8 in dfuse_open (pathname=0x7ffce049525a "/tmp/dfs_test/3dcb9152-1653-4111-8a07-261345887331/test_file", flags=0) at src/client/dfuse/il/int_posix.c:749
#11 0x0000000000401d55 in main ()`

Comment (mchaarawi) [reply src/common/fail_loc.c:188]: this is probably a multi-thread bug like forgetting to release a lock?
I can see in daos_debug_init and daos_init that the error is checked from daos_fail_init() so not sure what you are referring to.
from what i can tell your change introduces a leak. maybe someone else more familiar with this code path needs to review this instead of me because i don't understand how it doesn't.

Comment (ashleypittman) [reply src/common/fail_loc.c]: Should be fixed now.  It was a combination of d_fault_inject_fini() having a bug and the IL not checking the failure.

**Review by jolivier23 — APPROVED** [review]

**Comment by ashleypittman** [comment]
I've updated the title and first comment of this PR and it should now be ready for review.  This PR improves run-time and allows us to enable more testing, however if that testing is enabled them multiple errors are encountered which causes failures so enabling this new test and fixing these errors is being handled separately.

File: utils/node_local_test.py [hunk 6]
```diff
@@ -1662,18 +1662,290 @@ def test_pydaos_kv(server, conf):
...
+        fid = 1
+        max_count = 0
+        finished = False
+
+        # List of fids to re-run under valgrind.
+        to_rerun = []
+
+        fatal_errors = False
+
+        while not finished or active:
```
Comment (jolivier23) [inline utils/node_local_test.py:1889]: I personally think this is easier to read if you add parenthesis around "not finished" so one doesn't need to remember order of operations rules

**Review by jolivier23 — APPROVED** [review]

File: utils/node_local_test.py (hunk 6 above)
Comment (ashleypittman) [reply utils/node_local_test.py:1889]: I think pylint would warn if I tried that, but I'll address if I need to revise this.

**Review by daltonbohning — APPROVED** [review]
I had a few questions, but otherwise LGTM.

File: src/client/array/dc_array.c [hunk 7]
```diff
@@ -2308,10 +2304,8 @@ add_record(daos_handle_t oh, daos_handle_t th, struct set_size_props *props)
 	int			rc;
 
 	D_ALLOC_PTR(params);
-	if (params == NULL) {
-		D_ERROR("Failed memory allocation\n");
+	if (params == NULL)
 		return -DER_NOMEM;
-	}
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/array/dc_array.c:2314]: D_ALLOC already prints an error, in a common format.  These extra D_ERROR calls do not add anything.

File: src/client/dfuse/il/int_posix.c [hunk 8]
```diff
@@ -88,15 +110,21 @@ ioil_shrink(struct ioil_cont *cont)
 
 	if (cont->ioc_dfs != NULL) {
 		rc = dfs_umount(cont->ioc_dfs);
-		if (rc != 0)
+		if (rc != 0) {
 			D_ERROR("dfs_umount() failed, %d\n", rc);
+			return;
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/il/int_posix.c:115]: If the umount fails then the cont_close would also fail.  There's a weird thing going on with ioil_shrink however because it's called once on close, but also once again when the program is exiting so if we inject an error into dfs_umount() then it'll fail the first time, but the second time will work and daos_cont_close will as well.

The code as it was before would fail dfs_umount, then go into daos_cont_close() and through a bunch of needless errors.

File: utils/node_local_test.py [hunk 9]
```diff
@@ -655,8 +657,15 @@ def __init__(self, daos, conf, pool=None, container=None, path=None):
         self.valgrind_file = None
         self.container = container
         self.conf = conf
-        self.cores = None
+        # Detect the number of cores and do something sensible, if there are
+        # more than 32 on the node then use 12, otherwise use the whole node.
+        num_cores = len(os.sched_getaffinity(0))
+        if num_cores > 32:
+            self.cores = 12
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply utils/node_local_test.py:664]: As good as anything else.  It's mainly to constrain the number of cores in CI when running on whole nodes.  Some tuning for this is on the list at some point.

File: utils/node_local_test.py [hunk 10]
```diff
@@ -1662,18 +1662,290 @@ def test_pydaos_kv(server, conf):
...
+
+    def __str__(self):
+        res = "Fault injection test of '{}'\n".format(' '.join(self.cmd))
+        res += 'Fault injection location {}\n'.format(self.loc)
+        if self.vh:
+            res += 'Valgrind enabled for this test'
+        if self.returncode is None:
+            res += 'Process not completed'
+        else:
+            res += 'Returncode was {}'.format(self.returncode)
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply utils/node_local_test.py:1721]: L1718 might want to be a elif.  There's already a \n at the end on L1715.

File: utils/node_local_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply utils/node_local_test.py:664]: Ah, thank you!

File: utils/node_local_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply utils/node_local_test.py:1721]: Ah, I missed the `else`. Anyway, I think once you start running these tests and looking at the output you'll be able to see where it needs to be cleaned up. Certainly not worth another round of CI :)
